File: scripts/claude_cli_stream_probe/cli_cadence.py
"""Measure Claude Code CLI stdout arrival cadence.

Runs `claude -p ... --output-format stream-json` and timestamps every line
as it arrives at this Python reader. Then prints elapsed-ms per line so we
can see whether output dribbles (streaming) or arrives in one burst.
"""
import os, sys, time, subprocess, json, shutil

# Windows CreateProcess can't resolve a bare "claude"; use claude.cmd from PATH or the exe directly.
exe = shutil.which("claude.cmd") or shutil.which("claude")
if not exe:
    # fall back to the well-known nvm4w location
    exe = r"B:\nvm4w\nodejs\claude.cmd"
print(f"[probe] using {exe}", flush=True)

T0 = time.time()
proc = subprocess.Popen(
    [exe, "-p",
     "从1数到20,每个数字单独一行,只输出数字",
     "--output-format", "stream-json"],
    stdout=subprocess.PIPE, stderr=subprocess.PIPE,
    bufsize=1, text=True, encoding="utf-8",
    shell=True,
)
lines = []
try:
    while True:
        line = proc.stdout.readline()
        if not line:
            break
        el = int((time.time() - T0) * 1000)
        lines.append((el, line.rstrip("\n")))
        print(f"T+{el:>6}ms | {line.rstrip()[:180]}", flush=True)
finally:
    proc.wait(timeout=20)

# cadence summary
times = [t for t,_ in lines]
if len(times) >= 2:
    deltas = [times[i+1]-times[i] for i in range(len(times)-1)]
    print(f"\n[summary] lines={len(times)} first=T+{times[0]}ms last=T+{times[-1]}ms "
          f"span={times[-1]-times[0]}ms  max_gap={max(deltas)}ms  mean_gap={sum(deltas)//len(deltas)}ms")

File: scripts/claude_cli_stream_probe/cli_cadence.mjs
// cli_cadence.mjs — spawn `claude -p ... --output-format stream-json`, timestamp
// every stdout chunk as it arrives. Shows whether the CLI dribbles events
// (streams) or emits them in one burst (buffered/non-streaming).
import { spawn } from "node:child_process";

const exe = "claude.cmd"; // resolved via PATH (git-bash spawns .cmd through the shim)
const t0 = Date.now();
const log = (s) => console.log(s);

log(`[probe] spawn ${exe} @ T+0ms`);

const child = spawn(exe, [
  "-p",
  "从1数到20,每个数字单独一行,只输出数字",
  "--output-format", "stream-json",
  "--verbose",
], { stdio: ["ignore", "pipe", "inherit"], shell: true });

let firstAt = null, lastAt = null, chunks = 0, total = 0;
let buf = "";
child.stdout.on("data", (d) => {
  const at = Date.now() - t0;
  if (firstAt === null) firstAt = at;
  lastAt = at; chunks++; total += d.length;
  buf += d.toString("utf8");
  let idx;
  while ((idx = buf.indexOf("\n")) >= 0) {
    const line = buf.slice(0, idx); buf = buf.slice(idx + 1);
    if (line.trim()) log(`T+${String(at).padStart(6)}ms ch${String(chunks).padStart(3)} len=${String(d.length).padStart(4)} | ${line.slice(0,170)}`);
  }
});
child.on("exit", (code) => {
  log(`\n[summary] exit=${code} chunks=${chunks} totalBytes=${total} firstAt=T+${firstAt}ms lastAt=T+${lastAt}ms span=${lastAt-firstAt}ms`);
});
